fix(compiler): exclude empty styles from emitted metadata

crisbeto opened the pull request:
Excludes styles that resolve to empty strings from the emitted metadata so that they don't result in empty `<style>` tags at runtime.

Fixes #31191.

crisbeto commented on packages/compiler-cli/src/ngtsc/annotations/component/src/resources.ts:
```diff
@@ -447,6 +447,25 @@ export function transformDecoratorToInlineResources(
   }
 
+  // Filter out empty styles.
+  if (metadata.has('styles')) {
```
I'm not excited about this change, but I couldn't find a better way to exclude the empty styles from `setClassMetadata`. That being said, are we actually using the `styles` from `setClassMetadata` for anything? Looking a bit further up in the same function, there's some logic that transforms something like `styleUrls: ['a.css', 'b.css']` into `styles: ['a.css', 'b.css']` which doesn't make sense.

atscott replied:
From the discussion in the fw-compiler chat, can you add a comment where this is being done to explain what's going on there.

crisbeto replied:
It's being done a few lines up inside this function.

crisbeto replied:
So I went back to this after cleaning up my debugging code and it seems to be behaving as expected. I probably got confused with how it behaves while I was debugging. I'll mark this as resolved for now.

crisbeto commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/interface.ts:
```diff
@@ -30,7 +30,7 @@ export class PartialEvaluator {
       originatingFile: sourceFile,
       absoluteModuleName: null,
-      resolutionContext: sourceFile.fileName,
+      resolutionContext: sourceFile?.fileName,
```
This fixes an issue where some of the AST nodes in the `ngtsc` tests didn't have a `sourceFile`. I wasn't totally sure if it's a test setup problem.

atscott reviewed (dismissed):
reviewed-for: size-tracking

crisbeto commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -218,8 +218,16 @@ export function compileComponentFromMetadata(
-    const strings = styleValues.map(str => constantPool.getConstLiteral(o.literal(str)));
-    definitionMap.set('styles', o.literalArr(strings));
+    const strings = styleValues.reduce((result, str) => {
+      if (str.trim().length > 0) {
+        result.push(constantPool.getConstLiteral(o.literal(str)));
+      }
```
No, it's because we have to filter them out in two places: `defineComponent` and `setClassMetadata`. This is the logic that removes it from `setClassMetadata`.

jessicajaniuk reviewed (dismissed):
reviewed-for: size-tracking

clydin commented on packages/compiler-cli/src/ngtsc/annotations/component/src/resources.ts:
```diff
@@ -447,6 +447,25 @@ export function transformDecoratorToInlineResources(
   }
 
+  // Filter out empty styles.
+  if (metadata.has('styles')) {
```
The `styles` string array argument should already contain the external, inline, and template styles in string form. Filtering that array for empty strings and only setting the styles metadata field if the array contains elements should eliminate the need for the partial evaluator.
Something like the following maybe (untested):
```
metadata.delete('styleUrls');
metadata.delete('styles');
const actualStyles = styles.filter(el => el.trim().length > 0);
if (actualStyles.length > 0) {
    metadata.set(
        'styles',
        ts.factory.createArrayLiteralExpression(
            actualStyles.map(s => ts.factory.createStringLiteral(s))));
}
```

clydin commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -6919,6 +6919,70 @@ function allTests(os: string) {
+      `);
+
+        env.driveMain();
+        const jsContents = env.getContents('test.js').replace(/<link .*>/g, '');
```
This regex is greedy.  Currently i don't think it affects the test though.  Maybe `/<link .*?>/g` or `/<link [^>]*>/g` to be safe?

crisbeto committed:
fix(compiler): exclude empty styles from emitted metadata

Excludes styles that resolve to empty strings from the emitted metadata so that they don't result in empty `<style>` tags at runtime.

Fixes #31191.

crisbeto commented:
Merge assistance: this had all the necessary approvals, but they were reset after I rebased. I also had to change one of the size check files, because the size was reduced slightly.

jelbourn reviewed (approved):
LGTM

Reviewed-for: global-approvers

dylhunn commented:
This PR was merged into the repository by commit bb8d7091c64a785758ae347b1a9bcdeeed95897c.

dylhunn commented:
Sue to conflicts with 13.3.x, I only merged this into main. Let me know if you want to do a patch port!
